### ISSUE TITLE
Implement parrallel loop for spectopo

### ISSUE THREAD

**Issue by neuromechanist** [eeglabdevelopers/eeglab#572]
Hi,

Plotting the ~spectrograms~ spectral density plots (PSDs) is essential for experimenting with the data, cleaning methods, channel rejection, etc.
A couple of years ago, with rearranging variables on the `spectopo.m`'s main for-loop (~line 930 onward), I could implement a `parfor` to compute the ~spectrograms~ PSD for channels/components in parallel.

I wonder if there is an interest in implementing it here as well. The time-saving is ~3x, even for running on a laptop (my example, base model MacBook Pro 14", 2021), and would scale for HD-EEG, longer datasets, and higher core counts.

As an example, for a typical resting-state EEG under the HBN, the main spectopo for-loop takes ~6.5 sec to compute.
Parfor results are slightly varied as the first runs also have an overhead of assigning the memory. It starts with ~3.5 sec for the first run, 2.5 sec for the second run, and 1.5 sec for the third and ongoing runs.

As I plot the spectopo very frequently after each filtering or data cleaning, it is basically 4x the speed.

I also implemented a version with an argument pair `use-parallel`, which is `off` by default and can be turned on if the user wants.

I can create a branch and make a pull request to introduce the implementation, with or without the argument pair. Please let me know which option is preferable.

Thanks
Yahya

**Comment by arnodelorme** [comment]
Yes, a spectrogram plugin would be nice. The MATLAB spectrogram function is pretty powerful.

```matlab
EEG = pop_loadset('filename','eeglab_data.set','filepath','/System/Volumes/Data/data/matlab/eeglab/sample_data/');

[S,F,T,P] = spectrogram(EEG.data(1,:), EEG.srate,[], [], EEG.srate);
figure; imagesc(T, F, abs(S(1:EEG.srate/2,:))); set(gca, 'ydir', 'normal'); caxis([0 1000]);
```

If you overlay events, I think it would look great.

![Screen Shot 2022-12-15 at 9 21 19 AM](https://user-images.githubusercontent.com/1872705/207926750-792b8c89-8584-49da-9482-a416e1bb92da.png)

**Comment by arnodelorme** [comment]
Could be your first plugin :-)

**Comment by neuromechanist** [comment]
My bad for using the *spectrogram*. I misspoke, I should have said PSD plots, or the spectopo plots such as this one:
```Matlab
pop_spectopo(EEG, 1, [0 EEG.times(end)], 'EEG' ,'percent',100,'freq', [6 10 22], 'freqrange',[2 200],'electrodes','off');
```
<img width="563" alt="image" src="https://user-images.githubusercontent.com/44906843/207928105-dca7458c-7ee0-4491-be0f-a0dd4f2247d6.png">.

The spectrogram is also very useful; I would love to work on it, actually. But I believe it needs its own issue and discussion.

**Comment by arnodelorme** [comment]
I am not sure spectopo would benefit from parfor, mostly because the MATLAB spectral functions is already parallelized. If you do an FFT in MATLAB you will see your processor usage jump to 500% even for a single thread. Also parfor are a pain to handle in a robust way. One function that would be nice to program at the STUDY level is the tftopo function https://eeglab.org/tutorials/11_Scripting/EEG_scalp_measures.html

STUDY already use the parallel toolbox. Let's discuss more what can be done in EEGLAB and what overlaps with your interests.

**Comment by neuromechanist** [comment]
Absolutely,
thanks.
